development/cpp/custom_resource_format_loaders.rst

```diff
@@ -0,0 +1,260 @@
+Resources are primary containers. When load is called on the same file 
+path again, the previous loaded Resource will be referenced. Naturally, 
+loaded resources must be stateless.
+
+This guide assumes the reader knows how to create C++ modules and godot 
+data types. If not, refer to this guide :ref:`doc_custom_modules_in_c++`.
+
+
+References:
+~~~~~~~~~~~
+
+- :ref:`ResourceLoader<class_resourceloader>`
+- `core/io/resource_loader.cpp <https://github.com/godotengine/godot/blob/master/core/io/resource_loader.cpp#L258>`__
+
+What for?
+---------
+
+- Adding new support for many file formats
+- Audio format 
+- Video formats
+- ML Models
+
+What not?
+---------
+
+- Raster Images
+
+Developers should use ImageFormatLoader instead
+
+References:
+~~~~~~~~~~~
+
+- `core/io/image_loader.h <https://github.com/godotengine/godot/blob/master/core/io/image_loader.h>`__
+
+
+Creating a ResourceFormatLoader
+-------------------------------
+
+Each file format consist of a data container and a ``ResourceFormatLoader``. 
+
+ResourceFormatLoaders are usually simple classes which returns all the 
+necessary metadata for godot to understanding the new extension. The 
+class must the return format name and the extension string.
+
+In addition, ResourceFormatLoaders must convert file paths into 
+resources with the function ``load``. To load a resource, the ``load`` must 
+read and handle data serialization. 
+
+
+.. code:: cpp
+
+	#ifndef MY_JSON_LOADER_H
+	#define MY_JSON_LOADER_H
+
+	#include "io/resource_loader.h"
+
+	class ResourceFormatLoaderMyJson : public ResourceFormatLoader {
+	public:
+		virtual RES load(const String &p_path, const String &p_original_path, Error *r_error = NULL);
+		virtual void get_recognized_extensions(List<String> *p_extensions) const;
+		virtual bool handles_type(const String &p_type) const;
+		virtual String get_resource_type(const String &p_path) const;
+
+		ResourceFormatLoaderMyJson();
+		virtual ~ResourceFormatLoaderMyJson() {}
+	};
+	#endif // MY_JSON_LOADER_H
+
+.. code:: cpp
+
+	#include "my_json_loader.h"
+	#include "my_json.h"
+
+	ResourceFormatLoaderMyJson::ResourceFormatLoaderMyJson() {
+	}
+	RES ResourceFormatLoaderMyJson::load(const String &p_path, const String &p_original_path, Error *r_error) {
+		MyJson *my = memnew(MyJson);
+		if (r_error)
+			*r_error = OK;
+		Error err = my->set_file(p_path);
+		return Ref<MyJson>(my);
+	}
+
+	void ResourceFormatLoaderMyJson::get_recognized_extensions(List<String> *p_extensions) const {
+
+		p_extensions->push_back("mjson");
+	}
+	String ResourceFormatLoaderMyJson::get_resource_type(const String &p_path) const {
+
+		if (p_path.get_extension().to_lower() == "mjson")
+			return "MyJson";
+		return "";
+	}
+
+	bool ResourceFormatLoaderMyJson::handles_type(const String &p_type) const {
+		return (p_type == "MyJson");
+	}
+
+
+Creating Custom Data Types
+--------------------------
+
+
+Sometimes, Godot do not have the necessary facilities to express every data format. 
```

Comment:
do = does. Also... I don't really get what this says? Could you expand upon this?

Reply:
Godot does not have the proper data types for things like ML models. In order to expose ML models to GDscript, users must create custom classes,

Reply:
I am going to rewrite that part then. I was a bit too tired